File: app/src/main/java/io/github/hyuwah/muslimcompanionapp/Presenter/AyahFetcherPresenter.java
package io.github.hyuwah.muslimcompanionapp.Presenter;

import io.github.hyuwah.muslimcompanionapp.Contract.AyahFetcherContract;
import io.github.hyuwah.muslimcompanionapp.Model.Entity.Ayah;
import io.github.hyuwah.muslimcompanionapp.Model.Network.ServiceGenerator;
import io.github.hyuwah.muslimcompanionapp.Model.Network.AlquranCloudService;
import io.github.hyuwah.muslimcompanionapp.Model.SharedPrefsManager;
import io.github.hyuwah.muslimcompanionapp.Model.SharedPrefsManager.Key;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class AyahFetcherPresenter implements AyahFetcherContract.Presenter {

  private AyahFetcherContract.View view;

  public AyahFetcherPresenter(AyahFetcherContract.View mainView) {
    this.view = mainView;
  }

  @Override
  public void fetchRandomAyah() {
    int id = (int) Math.floor(Math.random() * ((6236 - 1 + 1) + 1)); // Generate id from 1 to 6236
    fetchAyah(id);
  }

  @Override
  public void fetchAyah(int id) {
    AlquranCloudService service = ServiceGenerator.createService(AlquranCloudService.class);

    // Check edition from prefs
    String edition = SharedPrefsManager.getInstance().getString(Key.EDITION_KEY, "en.asad");
    Call call = service.getAyahByEdition(id,edition);

    view.showLoading(); // Show loading

    call.enqueue(new Callback<Ayah>() {
      @Override
      public void onResponse(Call<Ayah> call, Response<Ayah> response) {
        view.fetchSuccess();
        if (response.body() != null) {
          view.showResult(response.body().getData());

          // save current ayah id to prefs
          SharedPrefsManager.getInstance().put(Key.CURRENT_AYAH_ID_INT, response.body().getData().getNumber());
        } else {
          view.fetchFailed();
        }
      }

      @Override
      public void onFailure(Call<Ayah> call, Throwable t) {
        view.fetchFailed();
      }
    });
  }

  @Override
  public void setAyahEdition(String edition) {
    SharedPrefsManager.getInstance().put(Key.EDITION_KEY, edition);
  }

  public void detachView() {
    view = null;
  }
}

File: app/src/main/java/io/github/hyuwah/muslimcompanionapp/Presenter/AyahFetcherPresenter.kt
package io.github.hyuwah.muslimcompanionapp.Presenter

import io.github.hyuwah.muslimcompanionapp.Contract.AyahFetcherContract
import io.github.hyuwah.muslimcompanionapp.Contract.AyahFetcherContract.Presenter
import io.github.hyuwah.muslimcompanionapp.Model.SharedPrefsManager
import io.github.hyuwah.muslimcompanionapp.domain.AlQuranCloudRepository
import kotlinx.coroutines.*
import org.koin.core.KoinComponent
import org.koin.core.inject

class AyahFetcherPresenter(private var view: AyahFetcherContract.View?) : KoinComponent, Presenter {

    private val repository: AlQuranCloudRepository by inject()
    private var job: Job? = null

    override fun fetchRandomAyah() {
        val id = Math.floor(Math.random() * (6236 - 1 + 1 + 1)).toInt() // Generate id from 1 to 6236
        fetchAyah(id)
    }

    override fun fetchAyah(id: Int) {
        val edition = SharedPrefsManager.getInstance().getString(SharedPrefsManager.Key.EDITION_KEY, "en.asad")
        view?.showLoading() // Show loading
        job = GlobalScope.launch(Dispatchers.IO) {
            val response = repository.getAyatByEdition(id, edition)
            withContext(Dispatchers.Main) {
                if (response.isSuccessful && response.body() != null) {
                    view?.showResult(response.body()!!.data)
                    // save current ayah id to prefs
                    SharedPrefsManager.getInstance().put(SharedPrefsManager.Key.CURRENT_AYAH_ID_INT, response.body()!!.data.number)
                } else {
                    view?.fetchFailed()
                }
            }
        }
    }

    override fun setAyahEdition(edition: String) {
        SharedPrefsManager.getInstance().put(SharedPrefsManager.Key.EDITION_KEY, edition)
    }

    fun detachView() {
        job?.cancel()
        job = null
        view = null
    }

}